lib/phoenix/pubsub/redis_server.ex

```diff
@@ -0,0 +1,176 @@
+defmodule Phoenix.PubSub.RedisServer do
+  use GenServer
+  require Logger
+
+  @moduledoc """
+  `Phoenix.PubSub` adapter for Redis
+
+  See `Phoenix.PubSub.Redis` for details and configuration options.
+  """
+
+  @defaults [host: "127.0.0.1", port: 6379, password: ""]
+
+  @max_connect_attemps 3   # 15s to establish connection
+  @reconnect_after_ms 5000
+  @redis_msg_vsn 1
+
+  @doc """
+  Starts the server
+  """
+  def start_link(opts) do
+    GenServer.start_link(__MODULE__, opts, name: Dict.fetch!(opts, :name))
+  end
+
+  @doc """
+  Initializes the server.
+
+  An initial connection establishment loop is entered. Once `:eredis_sub`
+  is started successfully, it handles reconnections automatically, so we
+  pass off reconnection handling once we find an initial connection.
+  """
+  def init(opts) do
+    server_name = Keyword.fetch!(opts, :name)
+    local_name  = Keyword.fetch!(opts, :local_name)
+    opts = Dict.merge(@defaults, opts)
+    opts = Dict.merge(opts, host: String.to_char_list(to_string(opts[:host])),
+                            password: String.to_char_list(to_string(opts[:password])))
+
+    Process.flag(:trap_exit, true)
+    send(self, :establish_conn)
+
+    {:ok, %{local_name: local_name,
+            namespace: redis_namespace(server_name),
+            eredis_sub_pid: nil,
+            eredis_pid: nil,
+            status: :disconnected,
+            node_ref: :erlang.make_ref,
+            reconnect_attemps: 0,
+            opts: opts}}
+  end
+
+  def handle_call({:subscribe, pid, topic, link}, _from, state) do
+    if link, do: Process.link(pid)
+    {:reply, GenServer.call(state.local_name, {:subscribe, pid, topic}), state}
+  end
+
+  def handle_call({:unsubscribe, pid, topic}, _from, state) do
+    {:reply, GenServer.call(state.local_name, {:unsubscribe, pid, topic}), state}
+  end
+
+  def handle_call({:subscribers, topic}, _from, state) do
+    {:reply, GenServer.call(state.local_name, {:subscribers, topic}), state}
+  end
+
+  def handle_call({:broadcast, from_pid, topic, msg}, _from, state) do
```

Comment:
It seems we no longer need the pool indeed, given the performance is good. If we are going to use the pool, we would need to spawn tasks here which could mess up ordering. Unless we add last_message_id then we could re-order them. So :+1 for no pool.

Question though: does the redis client (not the pubsub) automatically reconnects too?